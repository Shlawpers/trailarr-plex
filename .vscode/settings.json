{
    "explorer.fileNesting.enabled": true,
    "editor.stickyScroll.enabled": true,
    "notebook.stickyScroll.mode": "indented",
    "[python]": {
        "editor.defaultFormatter": "ms-python.black-formatter",
        "editor.formatOnSave": true,
    },
    "flake8.args": [
        "--ignore=W293,E203,E704",
        "--max-line-length=100",
    ],
    "python.testing.pytestArgs": [
        "."
    ],
    "python.testing.cwd": "${workspaceFolder}/backend",
    "python.testing.unittestEnabled": false,
    "python.testing.pytestEnabled": true,
    "todo-tree.tree.showBadges": true,
    "python.analysis.autoImportCompletions": true,
    "task.allowAutomaticTasks": "off",
    "task.autoDetect": "off",
    "typescript.tsc.autoDetect": "off",
    "yaml.schemas": {
        "https://squidfunk.github.io/mkdocs-material/schema.json": "mkdocs.yml"
    },
    "yaml.customTags": [
        "!ENV scalar",
        "!ENV sequence",
        "!relative scalar",
        "tag:yaml.org,2002:python/name:material.extensions.emoji.to_svg",
        "tag:yaml.org,2002:python/name:material.extensions.emoji.twemoji",
        "tag:yaml.org,2002:python/name:pymdownx.superfences.fence_code_format"
    ]
}