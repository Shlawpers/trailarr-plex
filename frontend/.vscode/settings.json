{
  "[html]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true,
    "editor.rulers": [140]
  },
  "[css]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true,
    "editor.rulers": [140]
  },
  "[scss]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true,
    "editor.rulers": [140]
  },
  "[typescript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true,
    "editor.rulers": [140],
    "typescript.tsdk": "node_modules\\typescript\\lib"
  }
}
