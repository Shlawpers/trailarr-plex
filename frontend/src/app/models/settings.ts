export interface Settings {
    api_key: string
    app_data_dir: string
    exclude_words: string
    version: string
    server_start_time: string
    timezone: string
    log_level: string
    monitor_enabled: boolean
    monitor_interval: number
    trailer_file_name: string
    trailer_folder_movie: boolean
    trailer_folder_series: boolean
    trailer_resolution: number
    trailer_audio_format: string
    trailer_video_format: string
    trailer_subtitles_enabled: boolean
    trailer_subtitles_format: string
    trailer_subtitles_language: string
    trailer_file_format: string
    trailer_embed_metadata: boolean
    trailer_remove_sponsorblocks: boolean
    trailer_web_optimized: boolean
    update_available: boolean
    wait_for_media: boolean
    yt_cookies_path: string
}

export interface ServerStats {
    trailers_downloaded: number
    movies_count: number
    movies_monitored: number
    series_count: number
    series_monitored: number
}