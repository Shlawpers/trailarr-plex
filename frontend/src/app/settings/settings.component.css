/* File: settings.component.css */
.settings-container {
  display: flex;
  flex-direction: column;
  /* min-height: 100%; */
  overflow: hidden;
  position: relative;
}

.settingsnav {
  position: sticky;
  top: 0;
  left: 0;
  right: 0;
  z-index: 99;
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: row;
  background-color: var(--color-surface-container-high);
  color: var(--color-on-surface);

  @media (width < 765px) {
    order: 2;
    justify-content: space-around;
    /* Make this work properly */
    position: fixed;
    top: auto;
    bottom: 78px;
  }
}

.settingsnav > * {
  flex: 1;
}

.settings-content {
  order: 2;
  overflow: hidden;
  padding: 0.5rem;
  position: relative; /* for the sticky update message in components */

  @media (width < 765px) {
    margin-bottom: 80px;
    order: 1;
  }
}

.setnav-btn {
  margin: 0;
  padding: 0.8rem;
  box-sizing: border-box;
  text-decoration: none;
  color: inherit;
  text-align: center;
  font-weight: 400;
  transition: 0.3s;
  cursor: pointer;
  border-radius: 0;
  border-bottom: 2px solid transparent;
}

.setnav-btn.active {
  background-color: var(--color-secondary-container); 
  opacity: 1;
  font-weight: 600;
  color: var(--color-on-secondary-container);
  border-bottom: 2px solid var(--color-primary);
}

.setnav-btn:not(.active):hover {
  background-color: var(--color-secondary-container); 
  opacity: 0.6;
  font-weight: 500;
  color: var(--color-on-secondary-container);
  border-bottom: 2px solid var(--color-primary);
}

/* @media (max-width: 765px) {
    .settings-container {
        grid-template-rows: 1fr auto;
    }

    .settingsnav {
      grid-row: 2;
      justify-content:space-around;
      top: auto;
      bottom: 0;
    }

    .settings-content {
      grid-row: 1;
    }
} */