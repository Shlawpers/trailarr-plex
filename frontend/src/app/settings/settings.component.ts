import {ChangeDetectionStrategy, Component} from '@angular/core';
import {RouterLink, RouterLinkActive, RouterOutlet} from '@angular/router';
import {RouteAbout, RouteConnections, RouteTrailer} from 'src/routing';

@Component({
  selector: 'app-settings',
  imports: [RouterLink, RouterLinkActive, RouterOutlet],
  templateUrl: './settings.component.html',
  styleUrl: './settings.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SettingsComponent {
  protected readonly RouteAbout = RouteAbout;
  protected readonly RouteConnections = RouteConnections;
  protected readonly RouteTrailer = RouteTrailer;
}
