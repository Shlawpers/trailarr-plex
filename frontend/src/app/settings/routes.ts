import {Route} from '@angular/router';
import {
  RouteAbout,
  RouteAdd,
  RouteConnections,
  RouteEdit,
  RouteParamConnectionId,
  RouteParamProfileId,
  RouteProfiles,
  RouteTrailer,
} from 'src/routing';
import {AboutComponent} from './about/about.component';
import {AddConnectionComponent} from './connections/add-connection/add-connection.component';
import {EditConnectionComponent} from './connections/edit-connection/edit-connection.component';
import {ShowConnectionsComponent} from './connections/show-connections/show-connections.component';
import {AddProfileComponent} from './profiles/add-profile/add-profile.component';
import {EditProfileComponent} from './profiles/edit-profile/edit-profile.component';
import {ShowProfilesComponent} from './profiles/show-profiles/show-profiles.component';
import {SettingsComponent} from './settings.component';
import {TrailerComponent} from './trailer/trailer.component';

export default [
  {
    path: '',
    loadComponent: () => SettingsComponent,
    children: [
      {path: RouteConnections, component: ShowConnectionsComponent},
      {path: `${RouteConnections}/${RouteAdd}`, component: AddConnectionComponent},
      {path: `${RouteConnections}/${RouteEdit}/:${RouteParamConnectionId}`, component: EditConnectionComponent},
      {path: RouteProfiles, component: ShowProfilesComponent},
      {path: `${RouteProfiles}/${RouteAdd}`, component: AddProfileComponent},
      {path: `${RouteProfiles}/${RouteEdit}/:${RouteParamProfileId}`, component: EditProfileComponent},
      {path: RouteTrailer, component: TrailerComponent},
      {path: RouteAbout, component: AboutComponent},
      {path: '**', redirectTo: RouteTrailer, pathMatch: 'prefix'},
    ],
  },
] as Route[];
