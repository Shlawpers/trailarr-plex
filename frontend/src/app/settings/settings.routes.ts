import {Routes} from '@angular/router';
import {AboutComponent} from './about/about.component';
import {AddConnectionComponent} from './connections/add-connection/add-connection.component';
import {ConnectionsComponent} from './connections/connections.component';
import {EditConnectionComponent} from './connections/edit-connection/edit-connection.component';
import {ShowConnectionsComponent} from './connections/show-connections/show-connections.component';
import {SettingsComponent} from './settings.component';
import {TrailerComponent} from './trailer/trailer.component';

export const settingsRoutes: Routes = [
  {
    path: '',
    component: SettingsComponent,
    children: [
      {path: '', redirectTo: 'trailer', pathMatch: 'full'},
      {
        path: 'connections',
        component: ConnectionsComponent,
        children: [
          {path: '', component: ShowConnectionsComponent},
          {path: 'add', component: AddConnectionComponent},
          {path: 'edit/:id', component: EditConnectionComponent},
        ],
      },
      {path: 'trailer', component: TrailerComponent},
      {path: 'about', component: AboutComponent},
    ],
  },
];
