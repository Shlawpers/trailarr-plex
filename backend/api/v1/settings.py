from fastapi import APIRouter

from api.v1.models import Settings, UpdateLogin, UpdateSetting
from api.v1 import authentication
from config.settings import app_settings
from core.base.database.manager.general import (
    GeneralDatabaseManager,
    ServerStats,
)

settings_router = APIRouter(prefix="/settings", tags=["Settings"])


@settings_router.get("/")
async def get_settings() -> Settings:
    return Settings(**app_settings.as_dict())


@settings_router.get("/stats")
async def get_stats() -> ServerStats:
    return GeneralDatabaseManager().get_stats()


@settings_router.put("/update")
async def update_setting(update: UpdateSetting) -> str:
    if not update.key:
        return "Error updating setting: Key is required"
    if update.value is None or update.value == "":
        return "Error updating setting: Value is required"
    if not hasattr(app_settings, update.key):
        msg = "Error updating setting: Invalid key"
        msg += (
            f" '{update.key}'! Valid values are"
            f" {app_settings.as_dict().keys()}"
        )
        return msg
    setattr(app_settings, update.key, update.value)
    _new_value = getattr(app_settings, update.key, None)
    _name = update.key.replace("_", " ").title()
    return f"Setting {_name} updated to {_new_value}"


@settings_router.put("/updatelogin")
async def update_login(login: UpdateLogin) -> str:
    # Current username and password are required
    if not login.current_password:
        return "Error updating login: Current password is required!"

    # Verify the current password
    if not authentication.verify_password(login.current_password):
        return "Error updating login: Current password is incorrect!"

    # New username and password are optional, but at least one is required
    if login.new_username:
        # If only the new username is provided, set it
        if not login.new_password:
            return authentication.set_username(login.new_username)
        else:
            # If both are provided, set both
            authentication.set_username(login.new_username)
            authentication.set_password(login.new_password)
            return "Username and password updated successfully"
    # If only the new password is provided, set it
    if login.new_password:
        return authentication.set_password(login.new_password)
    # If neither is provided, return an error
    return "Error updating credentials: None were provided!"
