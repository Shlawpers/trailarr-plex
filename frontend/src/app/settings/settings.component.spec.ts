import {MockBuilder, MockedComponentFixture, MockRender} from 'ng-mocks';
import {SettingsComponent} from './settings.component';

describe('SettingsComponent', () => {
  let fixture: MockedComponentFixture<SettingsComponent>;

  beforeEach(() => MockBuilder(SettingsComponent));

  beforeEach(() => {
    fixture = MockRender(SettingsComponent);
    fixture.detectChanges();
  });

  it('has instance', () => expect(fixture.componentInstance).toBeTruthy());

  it('renders', () => expect(fixture).toMatchSnapshot());
});
