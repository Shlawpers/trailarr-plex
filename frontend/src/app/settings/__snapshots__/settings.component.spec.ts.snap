// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsComponent renders 1`] = `
<mock-render>
  <app-settings>
    <div
      class="settingsnav"
    >
      <a
        class="setnav-btn"
        ng-reflect-router-link="connections"
        ng-reflect-router-link-active="active"
        routerlinkactive="active"
      >
        Connections
      </a>
      <a
        class="setnav-btn"
        ng-reflect-router-link="trailer"
        ng-reflect-router-link-active="active"
        routerlinkactive="active"
      >
        Trailer
      </a>
      <a
        class="setnav-btn"
        ng-reflect-router-link="about"
        ng-reflect-router-link-active="active"
        routerlinkactive="active"
      >
        About
      </a>
    </div>
    <div
      class="settings-container"
    >
      <div
        class="settings-content"
      >
        <router-outlet />
      </div>
    </div>
  </app-settings>
</mock-render>
`;
